arch/generic/crc32_braid_c.c

```diff
@@ -211,6 +207,539 @@ Z_INTERNAL uint32_t PREFIX(crc32_braid)(uint32_t crc, const uint8_t *buf, size_t
+        in25 = *((uint64_t*) (input + i + (24*8))) ^ in3 ^ in14 ^ in18 ^ bitbuffer[inoffset + 24];
+        in26 = *((uint64_t*) (input + i + (25*8))) ^ in4 ^ in15 ^ in19 ^ bitbuffer[inoffset + 25];
+        in27 = *((uint64_t*) (input + i + (26*8))) ^ in5 ^ in16 ^ in20 ^ bitbuffer[inoffset + 26];
+        in28 = *((uint64_t*) (input + i + (27*8))) ^ in6 ^ in17 ^ in21 ^ bitbuffer[inoffset + 27];
+        in29 = *((uint64_t*) (input + i + (28*8))) ^ in7 ^ in18 ^ in22 ^ bitbuffer[inoffset + 28];
+        in30 = *((uint64_t*) (input + i + (29*8))) ^ in8 ^ in19 ^ in23 ^ bitbuffer[inoffset + 29];
+        in31 = *((uint64_t*) (input + i + (30*8))) ^ in9 ^ in20 ^ in24 ^ bitbuffer[inoffset + 30];
+        in32 = *((uint64_t*) (input + i + (31*8))) ^ in10 ^ in21 ^ in25 ^ bitbuffer[inoffset + 31];
+
+        next1 = in11 ^ in22 ^ in26;
+        next2 = in12 ^ in23 ^ in27;
+        next3 = in13 ^ in24 ^ in28;
+        next4 = in14 ^ in25 ^ in29;
+        next5 = in15 ^ in26 ^ in30;
+        next6 = in16 ^ in27 ^ in31;
+        next7 = in17 ^ in28 ^ in32;
+        next8 = in18 ^ in29;
+        next9 = in19 ^ in30;
+        next10 = in20 ^ in31;
+        next11 = in21 ^ in32;
+        next12 = in22;
+        next13 = in23;
+        next14 = in24;
+        next15 = in25;
+        next16 = in26;
+        next17 = in27;
+        next18 = in28;
+        next19 = in29;
+        next20 = in30;
+        next21 = in31;
+        next22 = in32;
+
+        bitbuffer[outoffset1 + 22] = in1;
+        bitbuffer[outoffset1 + 23] = in2;
+        bitbuffer[outoffset1 + 24] = in3;
+        bitbuffer[outoffset1 + 25] = in4;
+        bitbuffer[outoffset1 + 26] = in5;
+        bitbuffer[outoffset1 + 27] = in6;
+        bitbuffer[outoffset1 + 28] = in7;
+        bitbuffer[outoffset1 + 29] = in8;
+        bitbuffer[outoffset1 + 30] = in9;
+        bitbuffer[outoffset1 + 31] = in10;
+        bitbuffer[outoffset2 + 0] = in11;
+        bitbuffer[outoffset2 + 1] = in12;
+        bitbuffer[outoffset2 + 2] = in13;
+        bitbuffer[outoffset2 + 3] = in14;
+        bitbuffer[outoffset2 + 4] = in15;
+        bitbuffer[outoffset2 + 5] = in16;
+        bitbuffer[outoffset2 + 6] = in17;
+        bitbuffer[outoffset2 + 7] = in18;
+        bitbuffer[outoffset2 + 8] = in19;
+        bitbuffer[outoffset2 + 9] = in20;
+        bitbuffer[outoffset2 + 10] = in21;
+        bitbuffer[outoffset2 + 11] = in22;
+        bitbuffer[outoffset2 + 12] = in23;
+        bitbuffer[outoffset2 + 13] = in24;
+        bitbuffer[outoffset2 + 14] = in25;
+        bitbuffer[outoffset2 + 15] = in26;
+        bitbuffer[outoffset2 + 16] = in27;
+        bitbuffer[outoffset2 + 17] = in28;
+        bitbuffer[outoffset2 + 18] = in29;
+        bitbuffer[outoffset2 + 19] = in30;
+        bitbuffer[outoffset2 + 20] = in31;
+        bitbuffer[outoffset2 + 21] = in32;
+    }
+
+    for(; i + 118960 + 512 < len; i += 256) {
+        uint64_t in1, in2, in3, in4, in5, in6, in7, in8;
```

Comment:
This is some pretty heavy unrolling. Is there anyway to maybe roll the loop body up and get the same performance?  I'm just thinking this is going to be generating a lot of program code and it doesn't seem like it needs this much to fill most pipelines.  Do you have slightly less unrolled variants we could test?

Reply:
If the scope of your variables are that long lived then rolling the loop up a bit more probably doesn't help that situation. I was thinking if this loop were rolled up a bit more the lifetimes of these variables would be short lived enough that they could simply be clobbered. Register moves on nearly every modern architecture go through move elimination and end up being almost free.

Reply:
If you could reduce things to something like 16 live variables to reduce stack spills on x86.that might be beneficial but your stashing values into a stack or heap backed buffer anyhow so you're ultimately going to be limited by that.

I'll be curious to try this on pre nehalem hardware that lack the carryless multiplies.

Reply:
yeah the stack spill is unavoidable unless someone has a clever trick, this is the densest polynomial i could find (so 22 is the minimum and then 32 total to avoid the DIV penalty)